### ISSUE TITLE
Make sure cursors are closed everywhere / Database is locked

### ISSUE THREAD

**Issue by PVince81** [owncloud/core#10566]
This follows my discovery here: https://github.com/owncloud/core/pull/7539#issuecomment-52849985

In some cases when prepared statements are used, if the cursors aren't closed explicitly, there is a higher risk to run into "Database is locked" issues on SQLite (and maybe max cursors reached on Oracle). We cannot rely on garbage collection to free these resources.

We should go through the core code and find the prepared statements that aren't closed explicitly.
The most likely to cause trouble is ones that might be called repeatedly or the ones defined as class attributes instead of local variables.

What do you think ?

@karlitschek @bantu @icewind1991 @DeepDiver1975 @th3fallen @butonic

**Comment by PVince81** [comment]
@butonic @DeepDiver1975 do you guys have an idea what happened with the "busyTimeout()" setting that we had in OC 7?

**Comment by PVince81** [comment]
Looks like we might be on the right track: https://github.com/owncloud/core/pull/13752

I can have a look at all "SELECT" statements in code and add an explicit "closeCursor()".
And we can make it a rule that all cursors must ALWAYS be closed explicitly.
This way we would avoid those special cases where closing might be important.

**Comment by PVince81** [comment]
Another case here: https://github.com/owncloud/core/pull/13772

We have this tendency of doing "check whether the value is in the DB with a SELECT" and then directly after that doing an "UPDATE" if the value is different. If the "UPDATE" is done in the same function, the `$result` isn't garbage collected and the cursor stays open.

SQLite hates that...

**Comment by oparoz** [comment]
@Raydiation - Is this automatic in the AppFramework?

**Comment by PVince81** [comment]
A few suspicious cases here, but I'm not 100% sure whether the GC would close the cursors there:
	- https://github.com/owncloud/core/blob/failed-delete-cache/apps/files_sharing/lib/share/file.php#L187
	- https://github.com/owncloud/core/blob/failed-delete-cache/apps/files_sharing/lib/share/folder.php#L76
	- https://github.com/owncloud/core/blob/master/apps/files_sharing/lib/helper.php#L171

**Comment by PVince81** [comment]
Another one:
https://github.com/owncloud/core/blob/master/apps/files_sharing/lib/external/manager.php#L114

Here we should probably first gather the results, close the cursor and only then proceed with mounting in case that takes longer.

**Comment by Raydiation** [comment]
@oparoz nope, not automatically done for you and I doubt that you can do it for all queries. It might be possible to close the cursor for findEntity and findEntities queries because all results are fetched, the same applies for insert and delete, but custom sql queries might executed with $mapper->execute() might need to fetch additional rows and you cant close the cursor there.

see https://github.com/owncloud/core/blob/master/lib/public/appframework/db/mapper.php

ideally we'd have something like a with statement, e.g.

```php
with($this->mapper->findAll() as $entities) {

}
```
where we could close the cursor when exiting from the block, but iirc php does not support this.

**Comment by oparoz** [comment]
Thank you @Raydiation. Maybe you or @PVince81 could add a small note to the documentation to make sure devs don't create more problems?

**Comment by PVince81** [comment]
Ideally we'd want to explicit close the cursor everywhere, but I'm having a look at existing queries. Because changing them all means need to retest everything, so I'm trying to find some suspicious ones that might need to be fixed earlier.

Having a look at this one I had doubt about the garbage collector, whether it would free the first result before the second query execution: https://github.com/owncloud/core/blob/master/apps/files_sharing/lib/share/folder.php#L76

Basically doing something like:
```php
$query = \OC_DB::prepare('SELECT ...');
$result = $query->execute(array('httpd/unix-directory'));
// ...
$query = \OC_DB::prepare('SELECT ...');
$result = $query->execute(array('httpd/unix-directory'));
```

Here the question is whether the second "$result" assignment will first destroy the old $result object (and implicitly call `closeCursor()`).

It turns out that the answer is no. It will first execute the second query by evaluating the right side, and only then destroy the old `$result` before assigning the new value.

This is proven by running this experiment: https://gist.github.com/PVince81/13ab190dcee79cbf5e5c
The output is as follows:
```
construct abc
construct def
destruct abc
destruct def
```

**Comment by PVince81** [comment]
@icewind1991 see comment above, we both had it wrong :wink:

**Comment by PVince81** [comment]
It also looks like the statement wrapper has no way to explicitly close the results: https://github.com/owncloud/core/blob/master/lib/private/db/statementwrapper.php#L72

But didn't we want to get rid of that wrapper anyway and consolidate the way how we do DB queries ?

**Comment by PVince81** [comment]
Places to look at:
        - [ ] close cursor in app framework: https://github.com/owncloud/core/pull/13777
	- [ ] https://github.com/owncloud/core/blob/failed-delete-cache/apps/files_sharing/lib/share/file.php#L187
	- [ ] https://github.com/owncloud/core/blob/failed-delete-cache/apps/files_sharing/lib/share/folder.php#L76
	- [ ] https://github.com/owncloud/core/blob/master/apps/files_sharing/lib/helper.php#L171
	- [ ] https://github.com/owncloud/core/blob/master/apps/files_sharing/lib/external/manager.php#L114
	- [ ] Search for ')->fetch'
	- [ ] https://github.com/owncloud/core/blob/master/apps/files_sharing/api/server2server.php#L162
	- [ ] https://github.com/owncloud/core/blob/master/apps/files_encryption/lib/util.php#L1369
	- [ ] https://github.com/owncloud/core/blob/master/lib/private/files/cache/cache.php#L409
	- [ ] file mapper: https://github.com/owncloud/core/blob/master/lib/private/files/mapper.php#L82
	- [ ] tags: https://github.com/owncloud/core/blob/master/lib/private/tags.php#L291
	- [ ] tags on user deletion: https://github.com/owncloud/core/blob/master/lib/private/tags.php#L517
	- [ ] unshare case: https://github.com/owncloud/core/blob/master/lib/private/share/helper.php#L116
	- [ ] sharing post_addToGroup: https://github.com/owncloud/core/blob/master/lib/private/share/hooks.php#L50
	- [ ] sharing getUsersSharingFile: https://github.com/owncloud/core/blob/master/lib/private/share/share.php#L126 (just check the whole share.php for SELECTs...)

**Comment by PVince81** [comment]
- [ ] repair mimetypes: https://github.com/owncloud/core/blob/master/lib/repair/repairmimetypes.php#L62
- [ ] a few more I fixed myself which might be more obvious: https://github.com/owncloud/core/pull/13778

**Comment by Raydiation** [comment]
@oparoz docs are updated

**Comment by oparoz** [comment]
Just went over the documentation, good job @Raydiation

**Comment by dragotin** [comment]
With ownCloud 8 RC1, set up with sqlite database. The system has three users: One user that shares with the two others. One of them creates 50 files with each size of 21 MiB:
```
for i in `seq 1 50`; do cp .xsession-errors-\:0 testpilotcloud/ordingaryfile$i; done
```

While uploading the files, the client reports an internal server error on one file, without further error messages in the server owncloud- or apache error log.

After the sync finished I removed a whole couple of these files again on the local side. This time, again one file threw an internal server error, this time with this error message:
```
[Mon Feb 02 14:07:55.142385 2015] [php5:error] [pid 3205] [client ::1:58152] PHP Fatal error:  Uncaught exception 'PDOException' with message 'SQLSTATE[HY000]: General error: 5 database is locked' in /home/kf/oC/core/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:91\nStack trace:\n#0 /home/kf/oC/core/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php(91): PDOStatement->execute(Array)\n#1 /home/kf/oC/core/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Statement.php(165): Doctrine\\DBAL\\Driver\\PDOStatement->execute(Array)\n#2 /home/kf/oC/core/lib/private/db/statementwrapper.php(63): Doctrine\\DBAL\\Statement->execute(Array)\n#3 /home/kf/oC/core/lib/private/db.php(141): OC_DB_StatementWrapper->execute(Array)\n#4 /home/kf/oC/core/lib/private/files/cache/cache.php(271): OC_DB::executeAudited('UPDATE `*PREFIX...', Array)\n#5 /home/kf/oC/core/lib/private/files/cache/changepropagator.php(63): OC\\Files\\Cache\\Cache->update('16', Array)\n#6 /home/kf/oC/core/lib/private/files/cache/updater.php(147): OC\\Files\\Cache\\ChangePropagator->propagateChanges()\n#7 [internal function]: OC\\Files\\Cache\\Updater->_ in /home/kf/oC/core/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractSQLiteDriver.php on line 84
```

**Comment by dragotin** [comment]
Ah, the good news is: In both cases the sync was "fixed" in the subsequent sync run that followed a couple of seconds after the errornous finished.

**Comment by DeepDiver1975** [comment]
The whole stacktrace above shows that this is the change propagation we perform at the end of a request execution. The execution path itself does not reveal any missing close cursors - no low hanging fruit from my pov.

@icewind1991 @PVince81 any ideas? THX

**Comment by PVince81** [comment]
@dragotin I don't see the "deadlock" pattern in that part of the code.
Note: the "deadlock" pattern is the one that is caused by unclosed
cursors that aren't garbage collected.

In your test case, did it all take more than one minute ?
I believe the default SQLite timeout is a minute, so we might want to
retest this together with the timeout setting from
https://github.com/owncloud/core/pull/13615 (unmerged) and see if it helps.

**Comment by dragotin** [comment]
@PVince81 nope, that was no timeout issue.

**Comment by PVince81** [comment]
From the stack trace it also looks like it was executed from inside the change propagator's `__destruct()` method which is automatically called on garbage collection, not sure from where. Maybe at that location there were still open cursors. CC @icewind1991

**Comment by dragotin** [comment]
Just removed around 30 files locally from the sync directory, the delete propagation to owncloud server post 8.0 release (SHA ac13cf04ba83d42b0bf6c2955e08de10634095c2) has internal server errors (client 1.8.0 beta1):
```
{"reqId":"2d126becb98ff1d94be833dddd33c5d4","remoteAddr":"::1","app":"remote","message":"An exception occurred while executing 'DELETE FROM \"oc_filecache\" WHERE \"fileid\" = ?' with params [793]:\n\nSQLSTATE[HY000]: General error: 5 database is locked","level":4,"time":"2015-02-17T15:43:26+00:00"}
{"reqId":"88d99c1c3500dc2b46bb920a80f063fa","remoteAddr":"::1","app":"remote","message":"An exception occurred while executing 'DELETE FROM \"oc_filecache\" WHERE \"fileid\" = ?' with params [815]:\n\nSQLSTATE[HY000]: General error: 5 database is locked","level":4,"time":"2015-02-17T15:43:30+00:00"}
{"reqId":"421e19a742f111026b6efd12b94cf0bc","remoteAddr":"::1","app":"remote","message":"An exception occurred while executing 'DELETE FROM \"oc_filecache\" WHERE \"fileid\" = ?' with params [922]:\n\nSQLSTATE[HY000]: General error: 5 database is locked","level":4,"time":"2015-02-17T15:43:39+00:00"}
{"reqId":"9ce133baaa3e7b273ee6cb82c8d5595a","remoteAddr":"::1","app":"remote","message":"An exception occurred while executing 'DELETE FROM \"oc_filecache\" WHERE \"fileid\" = ?' with params [942]:\n\nSQLSTATE[HY000]: General error: 5 database is locked","level":4,"time":"2015-02-17T15:43:40+00:00"}
{"reqId":"CvVy3ulsL1UkUd0sUi8r","remoteAddr":"::1","app":"remote","message":"An exception occurred while executing 'DELETE FROM \"oc_filecache\" WHERE \"fileid\" = ?' with params [952]:\n\nSQLSTATE[HY000]: General error: 5 database is locked","level":4,"time":"2015-02-17T15:46:20+00:00"}
```

**Comment by PVince81** [comment]
This seems to be the code for that query: https://github.com/owncloud/core/blob/master/lib/private/files/cache/cache.php#L370

Did you remove only files but also directories ? There seems to be a code path that would also delete child entries.

**Comment by PVince81** [comment]
The `$this->get()` does a SELECT on the same table and doesn't clear the cursor. But I'd expect that upon returning from the call, it should garbage collect the cursors automatically.

I also see that `$data` is actually the result of `fetchRow()` here: https://github.com/owncloud/core/blob/master/lib/private/files/cache/cache.php#L370 but I don't think that would cause the cursor to be kept.

That might be another argument to always explicitly close cursors and NEVER rely on the garbage collector.

@dragotin I get you aren't able to reproduce this consistently ? We could try and close the cursor there and see if it still happens.

**Comment by dragotin** [comment]
I removed only files in this case.

I can not really reproduce it, but it shows for me with good frequency when I remove a not too small amount of files locally (lets say 25).

**Comment by PVince81** [comment]
@dragotin does the client also delete in parallel by sending multiple HTTP requests ?

**Comment by PVince81** [comment]
Tagging this as junior job: there are many cases where it is easy to change the code to close the cursor.

PRs can be sent separately instead of a huge PR with everything, to make it easier to review and do regression testing on specific pieces of OC.

Complex cases can be left out to be tackled by non-juniors separately :smile:

**Comment by PVince81** [comment]
Had a chat with @dragotin and he found cases where we cannot rely on the garbage collector.
So the decision stays: we need to close the cursors everywhere.

**Comment by DeepDiver1975** [comment]
@karlitschek @cmonteroluque 

To late in the dev cycle - I suggest to postpone this to 9.0 or even backlog - THX

**Comment by karlitschek** [comment]
yes. too late for proper testing anyways

**Comment by cmonteroluque** [comment]
ok

**Comment by ownclouders** [comment]
Hey, this issue has been closed because the label `status/STALE` is set and there were no updates for 7 days. Feel free to reopen this issue if you deem it appropriate.

(This is an automated comment from [GitMate.io](https://gitmate.io).)

**Comment by PVince81** [comment]
Haven't heard of such issues any more, leaving closed.
